lib/rewards/server/message-actions.js

```diff
@@ -11,46 +11,77 @@ class RewardMessageActions
             return false;
         }
         if (ObjectsConst.OBJECT_INTERACTION !== message.act ||
-            RewardsConst.REWARDS_PICK_UP_ACT !== message.type) {
+            RewardsConst.REWARDS_PICK_UP_ACT !== message.type ||
+            !message['id']) {
             return false;
         }
-        let rewardId = message['id'];
-        await this.addItemToInventory(rewardId, room, playerSchema);
-        this.removeDrop(rewardId, room, client);
+
+        let rewardObject = this.getRewardObject(message['id'], room);
+        if (!this.isPlayerInsideDropInteractionArea(playerSchema, rewardObject, room)) {
+
+        }
+        await this.addItemToInventory(rewardObject.rewardModel, playerSchema);
+        this.removeDrop(rewardObject, room, client);
     }
 
-    removeDrop(rewardId, room, client)
+    removeDrop(rewardObject, room, client)
     {
-        let rewardObject = room.roomWorld.objectsManager.roomObjects[rewardId];
-        if (!rewardObject) {
-            return false;
-        }
         rewardObject.objectBody.world.removeBody(rewardObject.objectBody);
-        delete room.roomWorld.objectsManager.roomObjects[rewardId];
-        this.sendRemoveRewardMessage(rewardId, client);
+        let rewardObjectId = rewardObject.rewardModel.randomRewardId;
+        delete room.roomWorld.objectsManager.roomObjects[rewardObjectId];
+        this.sendRemoveRewardMessage(rewardObjectId, client);
     }
 
-    sendRemoveRewardMessage(rewardId, client)
+    sendRemoveRewardMessage(rewardObjectId, client)
     {
         // @TODO - BETA - Add new event here for the data.
         client.send('*', {
             act: RewardsConst.REWARDS_PICK_UP_ACT,
-            id: rewardId
+            id: rewardObjectId
         });
     }
 
-    async addItemToInventory(rewardId, room, playerSchema)
+    getRewardObject(rewardObjectId, room)
     {
-        let rewardObject = room.roomWorld.objectsManager.roomObjects[rewardId];
+        let rewardObject = room.roomWorld.objectsManager.roomObjects[rewardObjectId];
         if (!rewardObject) {
             return false;
         }
-        let rewardModel = rewardObject.rewardModel;
+        console.log('rewardObject found');
+        return rewardObject;
+    }
+
+    async addItemToInventory(rewardModel, playerSchema)
+    {
         let item = await playerSchema.skillsServer.dataServer.getEntity('item').loadById(rewardModel.itemId);
         let dropQty = rewardModel.dropQuantity;
         let rewardItem = playerSchema.inventory.manager.createItemInstance(item.key, dropQty);
         return await playerSchema.inventory.manager.addItem(rewardItem);
     }
+
+    isPlayerInsideDropInteractionArea(playerSchema, rewardObject, room)
+    {
+        // TODO: The interaction area should be defined on the rewardObject.
+        rewardObject.interactionArea = this.createInteractionAreaObject(room, rewardObject.x, rewardObject.y);
```

Comment:
this is because we mocked the roomObject with a normal object like `let roomObject = { // some properties }` right?
didn't checked the following comment, BUT: what if we create a RewardObject class that would extend a normal "AnimationObject" or "BaseObject" (or even just use the BaseObject), to get all the interaction stuff by default?

Reply:
that would save us from re adding the interaction area creation here, etc...

Reply:
exactly, if we extend from any of those two classes we would have the interaction area already on the room's object. The problem here is that those classes had a lot of properties and I don't know how to populate them (if that is needed).